js/packages/teams-ai/src/authentication/BotAuthentication.ts

```diff
@@ -0,0 +1,188 @@
+            },
+            true
+        );
+    }
+
+    public async authenticate(context: TurnContext, state: TState): Promise<string | undefined> {
+        // Get property names to use
+        const userAuthStatePropertyName = this.getUserAuthStatePropertyName(context);
+        const userDialogStatePropertyName = this.getUserDialogStatePropertyName(context);
+
+        // Save message if not signed in
+        if (!state.conversation.value[userAuthStatePropertyName]) {
+            state.conversation.value[userAuthStatePropertyName] = {
+                signedIn: false,
+                message: context.activity.text
+            };
+        }
+
+        const results = await this.runDialog(context, state, userDialogStatePropertyName);
+        if (results.status === DialogTurnStatus.complete && results.result != undefined) {
+            // Get user auth state
+            const userAuthState = state.conversation.value[userAuthStatePropertyName] as UserAuthState;
+            if (!userAuthState.signedIn && userAuthState.message) {
+                // Restore user message
+                context.activity.text = userAuthState.message;
+                userAuthState.signedIn = true;
+                delete userAuthState.message;
+                state.conversation.value[userAuthStatePropertyName] = userAuthState;
+            }
+
+            // Delete persisted dialog state
+            delete state.conversation.value[userDialogStatePropertyName];
+
+            // Return token
+            return results.result?.token;
+        } else {
+            return undefined;
+        }
+    }
+    /**
+     * The handler function is called when the user has successfully signed in
+     * @template TState
+     * @param {(context: TurnContext, state: TState) => Promise<void>} handler The handler function to call when the user has successfully signed in
+     */
+    public async onUserSignIn(handler: (context: TurnContext, state: TState) => Promise<void>): Promise<void> {
+        this._userSignInHandler = handler;
+    }
+
+    public async continueConversationalBotAuthFlow(context: TurnContext, state: TState): Promise<void> {
+        const userDialogStatePropertyName = this.getUserDialogStatePropertyName(context);
+        const result = await this.runDialog(context, state, userDialogStatePropertyName);
+
+        if (result.status === DialogTurnStatus.complete) {
+            if (result.result?.token) {
+                // Populate the token in the temp state
+                state.temp.value.authToken = result.result.token;
+
+                await context.sendActivity({
+                    value: { status: 200 } as InvokeResponse,
+                    type: ActivityTypes.InvokeResponse
+                });
+
+                // Successful sign in
+                await this._userSignInHandler?.(context, state);
+            } else {
+                await context.sendActivity({
```

Comment:
Seems OAuthPrompt has retry mechanism: https://github.com/microsoft/botbuilder-js/blob/main/libraries/botbuilder-dialogs/src/prompts/oauthPrompt.ts#L281. Will this impact how OAuthPrompt works?

Reply:
By the way, the [authentication sample](https://github.com/microsoft/BotBuilder-Samples/tree/main/samples/javascript_nodejs/18.bot-authentication) I saw previously does not return 400. If it's not necessary return a response, do you think we may remove this? The major purpose is keep things simple.

Reply:
The OAuthPrompt's retry mechanism is disabled by default. I throw an "AuthError" and let the application code decide whether to retry or not.

The only exception to this is that in the `botAuth.authenticate()` method if the dialog completes without a token, this means the user had cancelled the SSO flow after the previous turn. In this case we should clear dialog state and retry authentication: https://github.com/microsoft/teams-ai/pull/670/files#diff-5eec9e5c9f4146ed5353d5c5b52d82410317124563e16962d168a7a3d0d104d1R112-R116

Reply:
Good call @blackchoey, because the invoke response should be set from the OAuthPrompt dialog continue through the `runDialog()` call. Removed the invoke responses.